### ISSUE TITLE
Helm release of Kubernetes job times out despite the Job completing successfully

### ISSUE THREAD

**Issue by leiter-jakab** [helm/helm#9584]
Output of `helm version`:
version.BuildInfo{Version:"v3.5.3", GitCommit:"041ce5a2c17a58be0fcd5f5e16fb3e7e95fea622", GitTreeState:"dirty", GoVersion:"go1.15.8"}

Output of `kubectl version`:
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.5", GitCommit:"6b1d87acf3c8253c123756b9e61dac642678305f", GitTreeState:"clean", BuildDate:"2021-03-18T01:10:43Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.14", GitCommit:"5de7fd1f9555368a86eb0f8f664dc58055c17269", GitTreeState:"clean", BuildDate:"2021-01-18T09:31:01Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}

Cloud Provider/Platform (AKS, GKE, Minikube etc.): AKS

# Description

1. I have a helm chart that deploys a flyway migration as a kubernetes job. The chart is in local file system directory `./db-migration`.

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  labels:
    aadpodidbinding: key-vault-identity
  annotations:
    kubernetes.io/change-cause: {{ .Values.change_cause }}
  name: {{ .Values.application_name }}-database-migration-{{ .Values.version }}
spec:
  activeDeadlineSeconds: 420
  backoffLimit: 0
  template:
    metadata:
      labels:
        aadpodidbinding: key-vault-identity
    spec:
      containers:
        - args:
            - migrate
          env:
            - name: MIGRATION_USER
              valueFrom:
                secretKeyRef:
                  key: username
                  name: database-schema-admin
            - name: MIGRATION_PASSWORD
              valueFrom:
                secretKeyRef:
                  key: password
                  name: database-schema-admin
          image: "{{ .Values.environment }}{{ .Values.application_name | replace "-" "" }}.azurecr.io/{{ .Values.application_name }}-flyway-with-test-data:{{ .Values.version }}"
          imagePullPolicy: Always
          name: flyway-migration
          volumeMounts:
            - mountPath: /flyway/conf
              name: flyway-config
            - mountPath: /mnt/secrets-store
              name: secrets-store-inline
              readOnly: true
      restartPolicy: Never
      volumes:
        - configMap:
            name: {{ .Values.application_name }}-flyway-conf
          name: flyway-config
        - csi:
            driver: secrets-store.csi.k8s.io
            readOnly: true
            volumeAttributes:
              secretProviderClass: azure-kvname
          name: secrets-store-inline
```

2. I install the release with the following command:

```
helm upgrade --install db-migration-release db-migration \
    --namespace migration-namespace \
    --set version=1.0.0.0 \
    --set application_name=test \
    --set environment=test \
    --set change_cause=test \
    --wait --wait-for-jobs --cleanup-on-fail --timeout 7m10s
````

3. The migration runs to completion. The Job/Pod has status `Complete`.

# Expected result

The helm release finishes successfully shortly after the Job completes.

# Actual result

The helm release waits for the full timeout duration and finishes with failure due to timeout. Resources are not cleaned up after the failed release.

**Comment by bacongobbler** [comment]
Hi @leiter-jakab. It is unclear to me from the ticket whether the job succeeded in the time allotted or not. I understand that it entered the `Complete` state. When did that occur?

As far as resources not cleaning up after the release failing, there are several flags which can assist in that matter. But by default Helm leaves the cluster in that state so the user can step in and determine what caused the release to fail. Most notable example being the `--atomic` flag.

Can you be more specific about why the release failed? Perhaps some `--debug` logs may help determine why the release failed to run to completion. Perhaps it was waiting on other resources to become ready for example.

**Comment by leiter-jakab** [comment]
Hi @bacongobbler, sorry for the confusion.

> It is unclear to me from the ticket whether the job succeeded in the time allotted or not. I understand that it entered the Complete state. When did that occur?

The Kubernetes Job succeeded within the time allotted:
1. Job completed successfully.
2. Helm Release did not finish yet, but was running for the full timeout duration, after which it finished with a timout.

> As far as resources not cleaning up after the release failing, there are several flags which can assist in that matter. But by default Helm leaves the cluster in that state so the user can step in and determine what caused the release to fail. Most notable example being the --atomic flag.

I used the `--wait`, `--wait-for-jobs`, and `--cleanup-on-fail` flags.
1. Am I correct in my expectation that this should remove resources after a failed Release?
2. Is this equivalent to using `--atomic` in combination with `--wait-for-jobs`?

> Can you be more specific about why the release failed? Perhaps some --debug logs may help determine why the release failed to run to completion. Perhaps it was waiting on other resources to become ready for example.

Using the `--debug` flag I got the following logs:
```
history.go:56: [debug] getting history for release database-migration
install.go:173: [debug] Original chart version: ""
install.go:190: [debug] CHART PATH: /apps/database-migration

Release "database-migration" does not exist. Installing it now.
client.go:122: [debug] creating 2 resource(s)
wait.go:53: [debug] beginning wait for 2 resources with timeout of 2m0s
wait.go:194: [debug] Job is failed: moap-movie-database-migration/moap-movie-database-migration-1.0.0.3
.
.
.
wait.go:194: [debug] Job is failed: moap-movie-database-migration/moap-movie-database-migration-1.0.0.3
Error: timed out waiting for the condition
helm.go:81: [debug] timed out waiting for the condition
```

**Comment by leiter-jakab** [comment]
Additional info:
* As described the job runs to completion without errors, but the release fails with timeout.
* When I re-run the release with `ugrade --install` I see the following in the debug logs:
```
history.go:56: [debug] getting history for release database-migration
upgrade.go:123: [debug] preparing upgrade for database-migration
upgrade.go:131: [debug] performing update for database-migration
upgrade.go:303: [debug] creating upgraded release for database-migration
client.go:201: [debug] checking 2 resources for changes
client.go:464: [debug] Looks like there are no changes for ConfigMap "flyway-conf"
client.go:464: [debug] Looks like there are no changes for Job "database-migration-1.0.0.3"
wait.go:53: [debug] beginning wait for 2 resources with timeout of 40s
wait.go:194: [debug] Job is failed: database-migration/database-migration-1.0.0.3
.
.
.
upgrade.go:369: [debug] warning: Upgrade "database-migration" failed: timed out waiting for the condition
Error: UPGRADE FAILED: timed out waiting for the condition
helm.go:81: [debug] timed out waiting for the condition
UPGRADE FAILED
main.newUpgradeCmd.func2
	/home/circleci/helm.sh/helm/cmd/helm/upgrade.go:157
github.com/spf13/cobra.(*Command).execute
	/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850
github.com/spf13/cobra.(*Command).ExecuteC
	/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958
github.com/spf13/cobra.(*Command).Execute
	/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
main.main
	/home/circleci/helm.sh/helm/cmd/helm/helm.go:80
runtime.main
	/usr/local/go/src/runtime/proc.go:204
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1374
```
* The 2 resources (i.e. the Job and ConfigMap) are recognized as up by Helm as needing no changes. Despite this Helm goes into the status polling loop.

**Comment by dpp23** [comment]
I believe this is because of `BackoffLimit=0` (see https://github.com/helm/helm/issues/9637)

**Comment by bacongobbler** [comment]
bingo. Yes this is because the backofflimit is set to zero.

Closing as a duplicate of #9312 which will be fixe din Helm 3.6.
